benchmarks/opmphm.c

```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
+ */
+
+/**
+ * START ================================================== OPMPHM Search Time ======================================================= START
+ *
+ * This benchmark measures the time of the OPMPHM search.
+ * Uses all KeySet shapes, for one n (KeySet size) ksPerN KeySets are used.
+ * Each measurement done with one KeySet is repeated numberOfRepeats time and summarized with the median.
+ * For one n (KeySet size) the ksPerN results are also summarized with the median.
+ * The results are written out in the following format:
+ *
+ * n;search_1;search_2;...;search_(numberOfSearches)
+ *
+ * The number of needed seeds for this benchmarks is: numberOfShapes * nCount * ksPerN * (1  + searchesCount )
+ */
+
+/**
+ * @brief Measures the OPMPHM search time, for searches random Keys, repeats the measurement numberOfRepeats time and returns the media.
+ *
+ * The OPMPHM build will be triggerd if KDB_OPMPHM is set!
+ *
+ * @param ks the KeySet
+ * @param searches the number of searches to make
+ * @param searchSeed the random seed used to determine the Keys to search
+ * @param option the options passed to the ksLookup (...)
+ * @param repeats array to store repeated measurements
+ * @param numberOfRepeats fields in repeats
+ *
+ * @retval median time
+ */
+static size_t benchmarkSearchTimeMeasure (KeySet * ks, size_t searches, int32_t searchSeed, option_t option, size_t * repeats,
+					  size_t numberOfRepeats)
+{
+	if (option & KDB_O_OPMPHM)
+	{
+		// trigger OPMPHM build if not build
+		if (!ks->opmphm || !opmphmIsBuild (ks->opmphm))
+		{
+			// set seed to return by elektraRandGetInitSeed () in the lookup
+			elektraRandBenchmarkInitSeed = searchSeed;
+			(void) ksLookup (ks, ks->array[0], KDB_O_OPMPHM | KDB_O_NOCASCADING);
+			if (!ks->opmphm || !opmphmIsBuild (ks->opmphm))
+			{
+				printExit ("trigger OPMPHM build");
+			}
+		}
+	}
+	for (size_t repeatsI = 0; repeatsI < numberOfRepeats; ++repeatsI)
+	{
+		// sanity checks
+		if (option & KDB_O_OPMPHM)
+		{
+			if (!ks->opmphm || !opmphmIsBuild (ks->opmphm))
+			{
+				printExit ("Sanity Check Failed: OPMPHM not here");
+			}
+		}
+		else
+		{
+			if (ks->opmphm)
+			{
+				printExit ("Sanity Check Failed: OPMPHM here");
+			}
+		}
+
+		// preparation for measurement
+		struct timeval start;
+		struct timeval end;
+		Key * keyFound;
+		int32_t actualSearchSeed = searchSeed;
+
+		// START MEASUREMENT
+		__asm__("");
+		gettimeofday (&start, 0);
+		__asm__("");
+
+		for (size_t s = 1; s <= searches; ++s)
+		{
+			keyFound = ksLookup (ks, ks->array[actualSearchSeed % ks->size], option);
+			if (!keyFound || keyFound != ks->array[actualSearchSeed % ks->size])
+			{
+				printExit ("Sanity Check Failed: found wrong Key");
+			}
+			elektraRand (&actualSearchSeed);
+		}
+
+		__asm__("");
+		gettimeofday (&end, 0);
+		__asm__("");
+		// END MEASUREMENT
+
+		// sanity checks
+		if (option & KDB_O_OPMPHM)
```

Comment:
shouldn't that be in tests? (and not in benchmarks)

Reply:
Well I have such a check in the test cases, but I added them here to be sure to measure the right thing.